src/hamiltonian/hamiltonian_k.cpp

```diff
@@ -764,17 +763,26 @@ void Hamiltonian_k::set_fv_h_o_it(dmatrix<double_complex>& h__, dmatrix<double_c
 //==     }
 //== }
 
-template <typename T>
-void Hamiltonian_k::apply_h_s(spin_range spins__, int N__, int n__, Wave_functions<real_type<T>>& phi__,
-                               Wave_functions<real_type<T>>* hphi__, Wave_functions<real_type<T>>* sphi__)
+template <typename T> template <typename F, typename>
```

Comment:
What is T and what is F?

Reply:
> What is T and what is F?

T is the template type of Hamiltonian_k class (either float or double)
F is a further template type of the function that can real type or complex type while the precision type is same as T

Reply:
Isn't T enough? The gamma- or generic k-point case is handled by Q_operator and D_operator internally. I'll leave it for you to decide. In future, it might be a good idea to treat Gamma-point case explicitly with a template parameter (not by a real/complex data type)

Reply:
I just try to follow the code as much as original, which is templated against real and complex type as the function called the apply_non_local_d_q function which then called the inner function that depends on the type